ext/featurefmt/rpm/contentmanifest/updater.go

```diff
@@ -0,0 +1,170 @@
+package contentmanifest
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/quay/clair/v3/pkg/httputil"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	lastUpdateDate  = time.Date(1970, time.January, 1, 1, 1, 1, 1, time.UTC)
+	lastHeaderQuery = time.Date(1970, time.January, 1, 1, 1, 1, 1, time.UTC)
+
+	mappingFileMutex = sync.Mutex{}
+	updaterMutex     = sync.Mutex{}
+)
+
+// LocalUpdaterJob periodically updates mapping file and store it in local storage
+type LocalUpdaterJob struct {
+	LocalPath string
+	URL       string
+}
+
+// Get translate repositories into CPEs using a mapping file
+func (updater *LocalUpdaterJob) Get(repositories []string) ([]string, error) {
+	mappingFileMutex.Lock()
+	defer mappingFileMutex.Unlock()
+	f, err := os.Open(updater.LocalPath)
+	if err != nil {
+		return []string{}, err
+	}
+	defer f.Close()
+	fileContent, err := ioutil.ReadAll(f)
+	if err != nil {
+		return []string{}, err
+	}
+	mappingContent := MappingFile{}
+	err = json.Unmarshal(fileContent, &mappingContent)
+	if err != nil {
+		return []string{}, err
+	}
+	cpes := []string{}
+	for _, repo := range repositories {
+		if repoCPEs, ok := mappingContent.Data[repo]; ok {
+			for _, cpe := range repoCPEs.CPEs {
+				cpes = appendUnique(cpes, cpe)
+			}
+		} else {
+			log.WithField("repository", repo).Debug("The repository is not present in a mapping file")
+		}
+	}
+	return cpes, nil
+}
+
+// Update fetches mapping file using HTTP and store it locally in regular intervals
+func (updater *LocalUpdaterJob) Update() error {
+	updaterMutex.Lock()
+	defer updaterMutex.Unlock()
+	if updater.shouldBeUpdated() {
+		log.Info("The repo2cpe mapping has newer version. Updating...")
+		data, lastModified, err := updater.fetch()
```

Comment:
do you really want to hold this lock while your updater fetches?

Reply:
The mutex needs to be locked before updater checks whether mapping needs to be updated. If I would move lock bit lower multiple processes would check last updated date and all of them would fetch a mapping.